inc/class-segment-transformer.php

```diff
@@ -0,0 +1,272 @@
+		$mapping = [
+			'type' => $type . '|static',
+			'anonymousId' => 'endpoint.Id',
+			'messageId' => 'endpoint.RequestId',
+			'timestamp' => 'event_timestamp|' . __CLASS__ . '::format_date',
+			'userId' => 'endpoint.User.UserId',
+
+			'context' => [
+				'active' => 'endpoint.EndpointStatus',
+				'app' => [
+					'name' => '',
+					'build' => '',
+					'version' => 'endpoint.Demographics.AppVersion',
+				],
+				'campaign' => [
+					'name' => 'attributes.qv_utm_campaign',
+					'source' => 'attributes.qv_utm_source',
+					'medium' => 'attributes.qv_utm_medium',
+					'term' => 'attributes.qv_utm_term',
+					'content' => '',
+				],
+				'device' => [
+					'id' => '',
+					'advertisingId' => '',
+					'manufacturer' => 'endpoint.Attributes.DeviceMake.0',
+					'model' => 'endpoint.Attributes.DeviceModel.0',
+					'name' => '',
+					'type' => 'endpoint.Attributes.DeviceType.0',
+					'version' => '',
+				],
+				'ip' => '',
+				'library' => [
+					'name' => '',
+					'version' => '',
+				],
+				'locale' => 'endpoint.Demographic.Locale',
+				'location' => [
+					'country' => 'endpoint.Location.Country',
+					'city' => 'endpoint.Location.City',
+					'latitude' => 'endpoint.Location.Latitude',
+					'longitude' => 'endpoint.Location.Longitude',
+				],
+				'os' => [
+					'name' => 'endpoint.Demographic.Platform',
+					'version' => 'endpoint.Demographic.PlatformVersion',
+				],
+				'page' => [
+					'path' => '',
+					'referrer' => 'attributes.referrer',
+					'search' => 'attributes.search',
+					'title' => 'attributes.title',
+					'url' => 'attributes.url',
+					'keywords' => [],
+				],
+				'referrer' => [
+					'type' => '',
+					'name' => '',
+					'url' => '',
+					'link' => '',
+				],
+				'screen' => [
+					'density' => '',
+					'height' => '',
+					'width' => '',
+				],
+				'timezone' => 'endpoint.Demographic.Timezone',
+				'groupId' => '', // Groups are handled separately via `group` calls.
+				'traits' => [
+					'sessions' => 'endpoint.Metrics.sessions',
+					'pageViews' => 'endpoint.Metrics.pageViews',
+					'endpoint.Attributes',
+					// TODO: Do we have logged-in user attributes somewhere ?
+				],
+				'userAgent' => '',
+			],
+		];
+
+		if ( $type === 'identify' ) {
+			$type_mapping = array_merge( $mapping, [
+				'context' => null, // too big for an identify call. ?
+			] );
+		} elseif ( $type === 'page' ) {
+			$type_mapping = array_merge( $mapping, [
+				'name' => $mapping['context']['page']['title'],
+				'properties' => $mapping['context']['page'],
+			] );
+		} elseif ( $type === 'track' ) {
+			$type_mapping = array_merge( $mapping, [
+				'event' => 'event_type',
+				'properties' => [
+					'endpoint.Attributes', // TODO: This doesn't seem right, can it be more targetted ?
+				],
+			] );
+		} elseif ( $type === 'group' ) {
+			$type_mapping = array_merge( $mapping, [
+				'context' => null,
+				'messageId' => null,
+			] );
+		}
```

Comment:
Feels like a switch might be more suitable here, but also could these customisations be done in the specific methods e.g. `track()`, `page()` etc...?

Reply:
A `switch` might be more readable 👍 , will go for that. I preferred to encapsulate all the mapping definition in one function so any changes are more readable / contextual.